File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/alias/AccountFaucetPreparedVerifier.java
package org.hyperledger.iroha.android.alias;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.client.JsonEncoder;
import org.hyperledger.iroha.android.crypto.IrohaHash;
import org.hyperledger.iroha.android.model.Executable;
import org.hyperledger.iroha.android.model.FeePaymentIntent;
import org.hyperledger.iroha.android.model.InstructionBox;
import org.hyperledger.iroha.android.model.JsonValue;
import org.hyperledger.iroha.android.model.NetworkId;
import org.hyperledger.iroha.android.model.TransactionPayload;
import org.hyperledger.iroha.android.model.instructions.RegisterAccountWirePayloadEncoder;
import org.hyperledger.iroha.android.model.instructions.TransferWirePayloadEncoder;
import org.hyperledger.iroha.android.norito.NoritoException;
import org.hyperledger.iroha.android.norito.SignedTransactionEncoder;
import org.hyperledger.iroha.android.tx.SignedTransaction;
import org.hyperledger.iroha.android.tx.SignedTransactionHasher;

/** Fail-closed verification for first-release account-faucet prepare and submit results. */
public final class AccountFaucetPreparedVerifier {
  private AccountFaucetPreparedVerifier() {}

  /**
   * Authenticates one exact prepared faucet transaction against independent fee and faucet
   * policy expectations.
   */
  public static SignedTransaction requireValidPrepared(
      final AccountFaucetPreparedTransactionV1 prepared,
      final AccountFaucetClaimV1 claim,
      final TairaPublicResetMutationBindingV1 binding,
      final FeePaymentIntent expectedFeePayment,
      final AccountFaucetPolicyV1 policy,
      final NetworkId expectedNetworkId) {
    Objects.requireNonNull(prepared, "prepared");
    Objects.requireNonNull(claim, "claim");
    Objects.requireNonNull(binding, "binding");
    Objects.requireNonNull(expectedFeePayment, "expectedFeePayment");
    Objects.requireNonNull(policy, "policy");
    Objects.requireNonNull(expectedNetworkId, "expectedNetworkId");
    if (!prepared.binding().toJsonMap().equals(binding.toJsonMap())
        || !prepared.claim().toJsonMap().equals(claim.toJsonMap())) {
      throw new IllegalArgumentException(
          "prepared faucet envelope differs from the exact claim or binding");
    }
    if (!prepared.semanticHashHex().equals(claim.semanticHashHex())
        || !prepared.accountId().equals(claim.accountId())) {
      throw new IllegalArgumentException(
          "prepared faucet semantic identity differs from the exact claim");
    }
    if (!prepared.assetDefinitionId().equals(policy.assetDefinitionId())
        || !prepared.assetId().equals(policy.assetDefinitionId() + "#" + claim.accountId())
        || !prepared.amount().equals(policy.amount())) {
      throw new IllegalArgumentException(
          "prepared faucet asset definition, destination, or amount differs from trusted policy");
    }
    if (!expectedFeePayment.hasSamePayerAndGasBound(prepared.feePayment())) {
      throw new IllegalArgumentException(
          "prepared faucet fee intent changed payer, sponsor revision, or gas bound");
    }
    if (!AccountOnboardingReceiptVerifier.verifyAuthoritySignature(
        policy.faucetAuthority(),
        PreparedTransactionSignatureV1.digest(
            PreparedTransactionSignatureV1.faucetPrepared(prepared)),
        prepared.serverSignature())) {
      throw new IllegalArgumentException(
          "prepared faucet server signature is invalid for the trusted authority");
    }

    final byte[] wire =
        PreparedTransactionSignatureV1.decodeLowerHex(prepared.signedTransactionWireHex());
    if (!PreparedTransactionSignatureV1.hexLower(sha256(wire))
        .equals(prepared.signedTransactionWireSha256())) {
      throw new IllegalArgumentException(
          "prepared faucet wire SHA-256 differs from the envelope");
    }
    final SignedTransaction transaction;
    try {
      transaction = SignedTransactionEncoder.decodeVersioned(wire);
      if (!java.util.Arrays.equals(SignedTransactionEncoder.encodeVersioned(transaction), wire)) {
        throw new IllegalArgumentException(
            "prepared faucet wire is not canonical fixed-V1 SignedTransaction");
      }
    } catch (final NoritoException error) {
      throw new IllegalArgumentException("prepared faucet wire is invalid", error);
    }
    if (!SignedTransactionHasher.hashHex(transaction).equals(prepared.transactionHashHex())) {
      throw new IllegalArgumentException(
          "prepared faucet transaction hash differs from the envelope");
    }
    final TransactionPayload payload = SignedTransactionEncoder.decodeCanonicalPayload(transaction);
    if (!AccountOnboardingReceiptVerifier.verifyAuthoritySignature(
        payload.authority(),
        IrohaHash.prehash(transaction.encodedPayload()),
        transaction.signature())) {
      throw new IllegalArgumentException("prepared faucet transaction signature is invalid");
    }
    if (!payload.networkId().equals(expectedNetworkId)
        || !AccountOnboardingReceiptVerifier.sameAccountIdentity(
            payload.authority(), policy.faucetAuthority())) {
      throw new IllegalArgumentException(
          "prepared faucet transaction network or authority was substituted");
    }
    if (!payload.feePayment().equals(prepared.feePayment())) {
      throw new IllegalArgumentException(
          "prepared faucet fee intent differs from the signed transaction");
    }
    final Map<String, JsonValue> expectedMetadata = new LinkedHashMap<>();
    expectedMetadata.put(
        "taira_public_reset_binding",
        JsonValue.parse(JsonEncoder.encode(binding.toJsonMap())));
    expectedMetadata.put(
        "taira_prepared_operation",
        JsonValue.string(AccountFaucetPreparedTransactionV1.OPERATION));
    expectedMetadata.put(
        "taira_prepared_semantic_hash", JsonValue.string(prepared.semanticHashHex()));
    if (!payload.metadata().equals(expectedMetadata)) {
      throw new IllegalArgumentException(
          "prepared faucet transaction metadata differs from the envelope");
    }
    requireExactInstructions(payload.executable(), claim, policy);
    return transaction;
  }

  /** Reconciles only a policy-authenticated prepared faucet envelope and its exact submit hash. */
  public static PreparedTransactionSubmitResponseV1 requireValidSubmitResponse(
      final PreparedTransactionSubmitResponseV1 response,
      final AccountFaucetPreparedTransactionV1 prepared,
      final FeePaymentIntent expectedFeePayment,
      final AccountFaucetPolicyV1 policy,
      final NetworkId expectedNetworkId,
      final int httpStatus) {
    requireValidPrepared(
        prepared,
        prepared.claim(),
        prepared.binding(),
        expectedFeePayment,
        policy,
        expectedNetworkId);
    if (httpStatus != 200 && httpStatus != 202) {
      throw new IllegalArgumentException("prepared faucet submit requires HTTP 200 or 202");
    }
    if (!response.binding().toJsonMap().equals(prepared.binding().toJsonMap())
        || !response.operation().equals(prepared.operation())
        || !response.transactionHashHex().equals(prepared.transactionHashHex())) {
      throw new IllegalArgumentException(
          "prepared faucet submit response is not bound to the exact envelope");
    }
    if (httpStatus == 202 && response.outcome() != PreparedTransactionOutcomeV1.PENDING) {
      throw new IllegalArgumentException("HTTP 202 prepared faucet submit must remain Pending");
    }
    return response;
  }

  private static void requireExactInstructions(
      final Executable executable,
      final AccountFaucetClaimV1 claim,
      final AccountFaucetPolicyV1 policy) {
    if (!executable.isInstructions()) {
      throw new IllegalArgumentException(
          "prepared faucet transaction must contain a direct instruction sequence");
    }
    final InstructionBox transfer =
        TransferWirePayloadEncoder.encodeAssetTransfer(
            policy.assetDefinitionId() + "#" + policy.faucetAuthority(),
            policy.amount(),
            claim.accountId());
    final List<InstructionBox> instructions = executable.instructions();
    final boolean exactTransfer = instructions.size() == 1 && instructions.get(0).equals(transfer);
    final InstructionBox registration =
        RegisterAccountWirePayloadEncoder.encodeRegisterAccount(claim.accountId());
    final boolean exactRegisterThenTransfer =
        instructions.size() == 2
            && instructions.get(0).equals(registration)
            && instructions.get(1).equals(transfer);
    if (!exactTransfer && !exactRegisterThenTransfer) {
      throw new IllegalArgumentException(
          "prepared faucet transaction instructions differ from the exact claim and trusted policy");
    }
  }

  private static byte[] sha256(final byte[] value) {
    try {
      return MessageDigest.getInstance("SHA-256").digest(value);
    } catch (final NoSuchAlgorithmException impossible) {
      throw new IllegalStateException("SHA-256 is unavailable", impossible);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/alias/AccountFaucetPreparedVerifier.kt
package org.hyperledger.iroha.sdk.alias

import java.security.MessageDigest
import org.hyperledger.iroha.sdk.client.JsonEncoder
import org.hyperledger.iroha.sdk.core.model.Executable
import org.hyperledger.iroha.sdk.core.model.FeePaymentIntent
import org.hyperledger.iroha.sdk.core.model.JsonValue
import org.hyperledger.iroha.sdk.core.model.NetworkId
import org.hyperledger.iroha.sdk.core.model.instructions.RegisterAccountWirePayloadEncoder
import org.hyperledger.iroha.sdk.core.model.instructions.TransferWirePayloadEncoder
import org.hyperledger.iroha.sdk.crypto.IrohaHash
import org.hyperledger.iroha.sdk.tx.SignedTransaction
import org.hyperledger.iroha.sdk.tx.SignedTransactionHasher
import org.hyperledger.iroha.sdk.tx.norito.SignedTransactionEncoder
import org.hyperledger.iroha.sdk.tx.norito.TransactionPayloadAdapter

/** Fail-closed verification for first-release account-faucet prepare and submit results. */
object AccountFaucetPreparedVerifier {

    /**
     * Authenticates one exact prepared faucet transaction against independent fee and faucet
     * policy expectations.
     */
    @JvmStatic
    fun requireValidPrepared(
        prepared: AccountFaucetPreparedTransactionV1,
        claim: AccountFaucetClaimV1,
        binding: TairaPublicResetMutationBindingV1,
        expectedFeePayment: FeePaymentIntent,
        policy: AccountFaucetPolicyV1,
        expectedNetworkId: NetworkId,
    ): SignedTransaction {
        require(prepared.binding.toJsonMap() == binding.toJsonMap() && prepared.claim.toJsonMap() == claim.toJsonMap()) {
            "prepared faucet envelope differs from the exact claim or binding"
        }
        require(prepared.semanticHashHex == claim.semanticHashHex() && prepared.accountId == claim.accountId) {
            "prepared faucet semantic identity differs from the exact claim"
        }
        require(
            prepared.assetDefinitionId == policy.assetDefinitionId &&
                prepared.assetId == "${policy.assetDefinitionId}#${claim.accountId}" &&
                prepared.amount == policy.amount,
        ) {
            "prepared faucet asset definition, destination, or amount differs from trusted policy"
        }
        require(expectedFeePayment.hasSamePayerAndGasBound(prepared.feePayment)) {
            "prepared faucet fee intent changed payer, sponsor revision, or gas bound"
        }
        require(
            AccountOnboardingReceiptVerifier.verifyAuthoritySignature(
                policy.faucetAuthority,
                PreparedTransactionSignatureV1.digest(PreparedTransactionSignatureV1.faucetPrepared(prepared)),
                prepared.serverSignature,
            ),
        ) { "prepared faucet server signature is invalid for the trusted authority" }

        val wire = decodeLowerHex(prepared.signedTransactionWireHex)
        require(hexLower(MessageDigest.getInstance("SHA-256").digest(wire)) == prepared.signedTransactionWireSha256) {
            "prepared faucet wire SHA-256 differs from the envelope"
        }
        val transaction = try {
            SignedTransactionEncoder.decodeVersioned(wire).also { decoded ->
                require(SignedTransactionEncoder.encodeVersioned(decoded).contentEquals(wire)) {
                    "prepared faucet wire is not canonical fixed-V1 SignedTransaction"
                }
            }
        } catch (error: RuntimeException) {
            throw IllegalArgumentException("prepared faucet wire is invalid", error)
        }
        require(SignedTransactionHasher.hashHex(transaction) == prepared.transactionHashHex) {
            "prepared faucet transaction hash differs from the envelope"
        }
        val payload = TransactionPayloadAdapter.validateCanonicalPayloadBytes(transaction.encodedPayload())
        require(
            AccountOnboardingReceiptVerifier.verifyAuthoritySignature(
                payload.authority,
                IrohaHash.prehash(transaction.encodedPayload()),
                transaction.signature(),
            ),
        ) { "prepared faucet transaction signature is invalid" }
        require(
            payload.networkId == expectedNetworkId &&
                AccountOnboardingReceiptVerifier.sameAccountIdentity(payload.authority, policy.faucetAuthority),
        ) { "prepared faucet transaction network or authority was substituted" }
        require(payload.feePayment == prepared.feePayment) {
            "prepared faucet fee intent differs from the signed transaction"
        }
        val expectedMetadata = mapOf(
            "taira_public_reset_binding" to JsonValue.parse(JsonEncoder.encode(binding.toJsonMap())),
            "taira_prepared_operation" to JsonValue.string(AccountFaucetPreparedTransactionV1.OPERATION),
            "taira_prepared_semantic_hash" to JsonValue.string(prepared.semanticHashHex),
        )
        require(payload.metadata == expectedMetadata) {
            "prepared faucet transaction metadata differs from the envelope"
        }
        requireExactInstructions(payload.executable, claim, policy)
        return transaction
    }

    /** Reconciles only a policy-authenticated prepared faucet envelope and its exact submit hash. */
    @JvmStatic
    fun requireValidSubmitResponse(
        response: PreparedTransactionSubmitResponseV1,
        prepared: AccountFaucetPreparedTransactionV1,
        expectedFeePayment: FeePaymentIntent,
        policy: AccountFaucetPolicyV1,
        expectedNetworkId: NetworkId,
        httpStatus: Int,
    ): PreparedTransactionSubmitResponseV1 {
        requireValidPrepared(
            prepared,
            prepared.claim,
            prepared.binding,
            expectedFeePayment,
            policy,
            expectedNetworkId,
        )
        require(httpStatus == 200 || httpStatus == 202) {
            "prepared faucet submit requires HTTP 200 or 202"
        }
        require(
            response.binding.toJsonMap() == prepared.binding.toJsonMap() &&
                response.operation == prepared.operation &&
                response.transactionHashHex == prepared.transactionHashHex,
        ) { "prepared faucet submit response is not bound to the exact envelope" }
        require(httpStatus != 202 || response.outcome == PreparedTransactionOutcomeV1.PENDING) {
            "HTTP 202 prepared faucet submit must remain Pending"
        }
        return response
    }

    private fun requireExactInstructions(
        executable: Executable,
        claim: AccountFaucetClaimV1,
        policy: AccountFaucetPolicyV1,
    ) {
        require(executable is Executable.Instructions) {
            "prepared faucet transaction must contain a direct instruction sequence"
        }
        val transfer = TransferWirePayloadEncoder.encodeAssetTransfer(
            "${policy.assetDefinitionId}#${policy.faucetAuthority}",
            policy.amount,
            claim.accountId,
        )
        val withoutRegistration = listOf(transfer)
        val withRegistration = listOf(
            RegisterAccountWirePayloadEncoder.encodeRegisterAccount(claim.accountId),
            transfer,
        )
        require(executable.instructions == withoutRegistration || executable.instructions == withRegistration) {
            "prepared faucet transaction instructions differ from the exact claim and trusted policy"
        }
    }

    private fun hexLower(bytes: ByteArray): String {
        val digits = "0123456789abcdef"
        return buildString(bytes.size * 2) {
            bytes.forEach { byte ->
                val value = byte.toInt() and 0xff
                append(digits[value ushr 4])
                append(digits[value and 0x0f])
            }
        }
    }
}
